Automatic generation of the setup script in the install phase of the bundle package

vvolkl opened the pull request:
This PR 
* adds a method "k4_generate_setup_script" to Ilcsoftpackage (which is becoming a library for custom spack functions). This is similar to `spack load --sh package > setup.sh` but does not evaluate the current environment (see docstring)
* adds an install phase to key4hep-stack, to do previously manual steps for an installation:
    - run k4_generate_setup_script and create  a bash setup script in its prefix
    - which also (optionally) sets up gcc since it cannot be a direct dependency
    - create a symlink to the setup script

tmadlener commented on packages/key4hep-stack/package.py:
```diff
@@ -232,4 +241,38 @@ class Key4hepStack(BundlePackage):
               "which are therefore not supported." \
               "See https://root-forum.cern.ch/t/devtoolset-gcc-toolset-compatibility/38286")
 
+    def install(self, spec, prefix):
```
There are a lot of inconsistent indentation levels in this function

tmadlener commented on packages/Ilcsoftpackage/package.py:
```diff
@@ -7,6 +7,69 @@
+    modifications = env_mod.group_by_name()
+    new_env = {}
+    env_set_not_prepend = {}
+    for name, actions in sorted(modifications.items()):
```
Inconsistent indentation levels below. The for statements seem to have 4 (?), while the if statements only seem to get 2

tmadlener commented on packages/Ilcsoftpackage/package.py:
```diff
@@ -7,6 +7,69 @@
+            else:
+              env_set_not_prepend[name] = False
+            # third, actually set an environment
+            x.execute(new_env)
```
Can you elaborate a bit why we also need to set the environment in this case? I would have naively assumed that this is not necessary when we only try to generate a setup script. Or does this simply fill the `new_env` variable?

vvolkl replied:
Yes, the "environment" in this case is just a dictionary that is filled here. updated the slightly misleading comment.

tmadlener commented on packages/Ilcsoftpackage/package.py:
```diff
@@ -7,6 +7,69 @@
+    k4_shell_prepend_strings = {
+        'sh': 'export {0}={1}:${0};\n',
+    }
+    cmds = ''
```
Can this be made a list to which the commands are appended and then in the end `return ''.join(cmds)`. This might help to avoid having to allocate extremely long strings over and over again.

vvolkl replied:
Sure, good point.

tmadlener commented on packages/key4hep-stack/package.py:
```diff
@@ -232,4 +241,38 @@ class Key4hepStack(BundlePackage):
+                        _spec.traverse( order='post')]
+              
+               if "+load_gcc" in spec:
+                 gcc_specs = [spack.cmd.disambiguate_spec("gcc", None, first=True)]
```
Will this always pick up the gcc version that has been used to build the `key4hep-stack`?

vvolkl replied:
No, it's true that this could be improved by parsing the spec. Having several gcc installations around is not a likely scenario though - this is really geared towards automating the workflow on build machines.

tmadlener reviewed (commented):
Have you already using this for generating a setup script for the nightly builds? Or for the cvmfs installation?

From what it looks like it is possible that some paths are in the final `PATH` more than one time, since they are not deduplicated. I think this should in general be fine, since any but the first occurence will probably be ignored for `PATH`. However, maybe we should keep an eye on this and check whether duplicated entries can lead to problems for other environment variables.

In general this looks pretty good to me.

vvolkl commented:
Hi Thomas,

thanks for the comments!

> Have you already using this for generating a setup script for the nightly builds? Or for the cvmfs installation?

Just tried it locally. But it should be picked up as soon as this is merged.

> From what it looks like it is possible that some paths are in the final `PATH` more than one time, since they are not deduplicated. I think this should in general be fine, since any but the first occurence will probably be ignored for `PATH`. However, maybe we should keep an eye on this and check whether duplicated entries can lead to problems for other environment variables.

Yes, I put it as a TODO. I don't think there are any side effects of having duplicate paths in the usual environment variables like PATH, LD_LIBRARY_PATH, ... there is a slight overhead though as packages like gaudi go through each of the paths to look for components.

vvolkl commented on packages/Ilcsoftpackage/package.py:
```diff
@@ -7,6 +7,69 @@
+    for name, actions in sorted(modifications.items()):
+        for x in actions:
+            if isinstance(x, SetPath) or isinstance(x, SetEnv):
+                env_set_not_prepend[name] = True
```
That's an interesting point. The assumption is that if there is a `SetEnv` action for a path, there should only be this one action. I'd consider it a bug of the package recipe if it tries to append to or set a path that has already been set. (Some packages like pythia8 also don't parse `:` in environment variables, otherwise I could have just skipped this part of the logic) But maybe there is a way to catch this error if it occurs.

andresailer replied:
There was a mistake in my comment. It will take the value of the last action, depending if this is a specific instance or not, so if there are multiple actions it just takes the last one. Though I don't know what modifications or actions are anyway :)

vvolkl commented:
I think I now adressed everything:

* added the deduplications for the paths
* parse the key4hep-stack for the compiler
* use exception handling instead of variants 
* added a warning if there is more than one action for a variable that has a `SetEnv` action associated.

vvolkl commented:
I'll merge in order to test this in tonight's nightlies, but feel free to continue to comment.
